dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPTX.java

```diff
@@ -33,6 +35,30 @@ public List<Event> visitPtxAtomOp(PTXAtomOp e) {
         );
     }
 
+    // PTX CAS semantics from
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-atom
+    // atomic {
+    //   d = *a;
+    //   *a = (*a == b) ? c : *a
+    // }
+    @Override
+    public List<Event> visitPtxAtomCAS(PTXAtomCAS e) {
+        Register resultRegister = e.getResultRegister();
+        String mo = e.getMo();
+        Expression address = e.getAddress();
+        Expression expected = e.getExpectedValue();
+        Expression newValue = e.getStoreValue();
+        Expression storeValue = expressions.makeConditional(expressions.makeBinary(resultRegister, COpBin.EQ, expected),
+                newValue, resultRegister);
```

Comment:
You can use the `makeEQ` method (not sure about the name). Also, this code does not align with the pseudo code, but from looking in the PTX docu, I think their pseudo code is misleading.
The pseudo code suggests that there is a fresh load `*a` happening instead of reusing the value in `d`. But it surely makes more sense to reuse the already loaded value.

Reply:
I don't think we should precisely follow the pseudo code. I just added it because the PTX CAS is slightly different than the other ones we have in the sense that it always performs the Store (it the comparison fails, it simple stores the old value).

Reply:
Are you sure that the ternary operator is meant as a `select/cmov` or as a proper branching? This makes a difference in dependencies.

Reply:
The documentation of atomic operations (which includes `cas`) states
```
Atomically loads the original value at location a into destination register d, performs a reduction operation with operand b and the value in location a, and stores the result of the specified operation at location a, overwriting the original value.
```
As I understand this, the store **is always executed**. 
If that is the case, branching makes nonsense to me.

Reply:
```
d = *a;
if (d == b)
  *a = c;
else 
  *a = d;
```
This is still feasible code that always stores. Though I agree that the code is most likely referring to a `cmov` (I just wanted to be sure).